File: AlgorithmsProblemSolvingJava/src/io/programminglife/algoexpert/medium/ReconstructBST.java
package io.programminglife.algoexpert.medium;

import java.util.ArrayList;

public class ReconstructBST {

    static class BST {
        public int value;
        public BST left = null;
        public BST right = null;

        public BST(int value) {
            this.value = value;
        }
    }

    static class TreeInfo {
        public int rootIdx;

        public TreeInfo(int rootIndex) {
            this.rootIdx = rootIndex;
        }
    }

    public BST reconstructBst(ArrayList<Integer> preOrderTraversalValues) {
        TreeInfo treeInfo = new TreeInfo(0);
        return reconstructBstRec(Integer.MIN_VALUE, Integer.MAX_VALUE, preOrderTraversalValues, treeInfo);
    }

    public BST reconstructBstRec(int lowerBound, int upperBound, ArrayList<Integer> preOrderTraversalValues,
            TreeInfo treeInfo) {
        if (preOrderTraversalValues.size() == treeInfo.rootIdx) {
            return null;
        }

        int rootValue = preOrderTraversalValues.get(treeInfo.rootIdx);
        if (rootValue < lowerBound || rootValue >= upperBound) {
            return null;
        }

        treeInfo.rootIdx += 1;
        BST leftSubtree = reconstructBstRec(lowerBound, rootValue, preOrderTraversalValues, treeInfo);
        BST rightSubtree = reconstructBstRec(rootValue, upperBound, preOrderTraversalValues, treeInfo);

        BST root = new BST(rootValue);
        root.left = leftSubtree;
        root.right = rightSubtree;

        return root;
    }

}

File: AlgorithmsProblemSolvingPython/algoexpert/medium/reconstructBst.py
class BST:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right


class TreeInfo:
    def __init__(self, rootIdx):
        self.rootIdx = rootIdx


def reconstructBst(preOrderTraversalValues):
    tree_info = TreeInfo(0)
    return reconstructBstRec(float("-inf"), float("inf"), preOrderTraversalValues, tree_info)


def reconstructBstRec(lowerBound, upperBound, preOrderTraversalValues, tree_info):
    if (len(preOrderTraversalValues) == tree_info.rootIdx):
        return None

    root_value = preOrderTraversalValues[tree_info.rootIdx]
    if (root_value < lowerBound or root_value >= upperBound):
        return None

    tree_info.rootIdx += 1
    left_tree = reconstructBstRec(
        lowerBound, root_value, preOrderTraversalValues, tree_info)
    right_tree = reconstructBstRec(
        root_value, upperBound, preOrderTraversalValues, tree_info)

    root = BST(root_value, left_tree, right_tree)

    return root
